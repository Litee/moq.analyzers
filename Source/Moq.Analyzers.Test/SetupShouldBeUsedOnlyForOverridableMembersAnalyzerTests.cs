using Verifier = Moq.Analyzers.Test.Helpers.AnalyzerVerifier<Moq.Analyzers.SetupShouldBeUsedOnlyForOverridableMembersAnalyzer>;

namespace Moq.Analyzers.Test;

public class SetupShouldBeUsedOnlyForOverridableMembersAnalyzerTests
{
    public static IEnumerable<object[]> TestData()
    {
        foreach (string @namespace in new[] { string.Empty, "namespace MyNamespace;" })
        {
            yield return [@namespace, """new Mock<BaseSampleClass>().Setup(x => {|Moq1200:x.Calculate()|});"""];
            yield return [@namespace, """new Mock<SampleClass>().Setup(x => {|Moq1200:x.Property|});"""];
            yield return [@namespace, """new Mock<SampleClass>().Setup(x => {|Moq1200:x.Calculate(It.IsAny<int>(), It.IsAny<int>(), It.IsAny<int>())|});"""];
            yield return [@namespace, """new Mock<BaseSampleClass>().Setup(x => x.Calculate(It.IsAny<int>(), It.IsAny<int>()));"""];
            yield return [@namespace, """new Mock<ISampleInterface>().Setup(x => x.Calculate(It.IsAny<int>(), It.IsAny<int>()));"""];
            yield return [@namespace, """new Mock<ISampleInterface>().Setup(x => x.TestProperty);"""];
            yield return [@namespace, """new Mock<SampleClass>().Setup(x => x.Calculate(It.IsAny<int>(), It.IsAny<int>()));"""];
            yield return [@namespace, """new Mock<SampleClass>().Setup(x => x.DoSth());"""];
        }
    }

    [Theory]
    [MemberData(nameof(TestData))]
    public async Task ShouldAnalyzeSetupForOverridableMembers(string @namespace, string mock)
    {
        await Verifier.VerifyAnalyzerAsync(
                $$"""
                {{@namespace}}

                public interface ISampleInterface
                {
                    int Calculate(int a, int b);
                    int TestProperty { get; set; }
                }

                public abstract class BaseSampleClass
                {
                    public int Calculate() => 0;
                    public abstract int Calculate(int a, int b);
                    public abstract int Calculate(int a, int b, int c);
                }

                public class SampleClass : BaseSampleClass
                {
                    public override int Calculate(int a, int b) => 0;
                    public sealed override int Calculate(int a, int b, int c) => 0;
                    public virtual int DoSth() => 0;
                    public int Property { get; set; }
                }

                internal class UnitTest
                {
                    private void Test()
                    {
                        {{mock}}
                    }
                }
                """);
    }
}
