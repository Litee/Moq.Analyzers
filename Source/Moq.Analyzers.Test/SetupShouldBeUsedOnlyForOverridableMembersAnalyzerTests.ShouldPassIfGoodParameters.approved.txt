Diagnostic 1
	Id: Moq1200
	Location: SourceFile(Test0.cs[2449..2462))
	Highlight: x.Calculate()
	Lines: mock.Setup(x => x.Calculate());
	Severity: Error
	Message: Setup should be used only for overridable members.

Diagnostic 2
	Id: Moq1200
	Location: SourceFile(Test0.cs[2619..2681))
	Highlight: x.Calculate(It.IsAny<int>(), It.IsAny<int>(), It.IsAny<int>())
	Lines: mock.Setup(x => x.Calculate(It.IsAny<int>(), It.IsAny<int>(), It.IsAny<int>()));
	Severity: Error
	Message: Setup should be used only for overridable members.

Diagnostic 3
	Id: Moq1200
	Location: SourceFile(Test0.cs[2844..2854))
	Highlight: x.Property
	Lines: mock.Setup(x => x.Property);
	Severity: Error
	Message: Setup should be used only for overridable members.

